### PR TITLE
TensorBoard Integration with Notebook Example

### PR DESCRIPTION
**Pull request by closedLoop**
Integration of TensorBoard with Notebook Example

@dustinvtran here's a quick demo of the changes
https://www.useloom.com/share/29e087b17cfc47b9b60662bff7843983

Contents:
 * Modification of `inference/inference.py` to more completely initialize Tensorboard
 * A Notebook with instructions and best practices to make your TensorBoard clean and useful

Main Features:
 * Automatically tracks all named latent variables for your given inference
   * This can be disabled or modified by passing in a custom list via `logvars`
 * Works for all inference classes.
 * Automatic run versioning to keep your logs clean and comparable in TensorBoard
 * Instructions for people to easily configure and launch their model with TensorBoard

Issues addressed
 * #464 satisfies part of the 'More TensorBoard support'
 * #190 the Notebook shows best practices for TensorBoard to look clean using Edward
 * #351 Partially addresses this by automatically visualizing all named latent variables
 * #353 This shows parameter MCMC values, but not loss functions.

Not addressed:
 * Storing inference specific loss and convergence information.  I saw that other were working on this and so I didn't touch it.  #598

### CODE REVIEW

File: edward/inferences/inference.py [hunk 1]
```diff
@@ -169,11 +171,25 @@ def initialize(self, n_iter=1000, n_print=None, scale=None, logdir=None,
     logdir : str, optional
       Directory where event file will be written. For details,
       see ``tf.summary.FileWriter``. Default is to write nothing.
+    logrun : str, optional
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: Could you describe your reasoning for splitting logdir vs logrun? I think I see it, although I wonder if logdir could just be made more powerful by letting you include a custom subdirectory naming. If unspecified, it would default to a subdirecty with the current UTC timestamp.

File: edward/inferences/inference.py [hunk 2]
```diff
@@ -194,6 +210,16 @@ def initialize(self, n_iter=1000, n_print=None, scale=None, logdir=None,
 
     if logdir is not None:
       self.logging = True
+
+      if logrun is None:
+        # Set default to timestamp
+        logrun = datetime.strftime(datetime.utcnow(), "%Y%m%dT%H%M%S")
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: Good standard! What's the T? 

Also, do you think adding `_` inbetween would make the directory name too long?

File: edward/inferences/inference.py [hunk 3]
```diff
@@ -258,3 +284,48 @@ def finalize(self):
     """
     if self.logging:
       self.train_writer.close()
+
+  def set_log_variables(self, logvars=None, log_max_scalers_per_var=None):
+    """Logs variables to TensorBoard
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: `Tensorboard.`

File: edward/inferences/inference.py [hunk 4]
```diff
@@ -258,3 +284,48 @@ def finalize(self):
...
+
+     For each variable in logvars, creates 'scalar' and / or 'histogram' by
+     calling `tf.summary.scalar` or `tf.summary.histogram`
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: To format code, wrap double single ticks, i.e., `\``tf.summary.scalar``\` (without the `\`). Also needs a period at end of sentence.

File: edward/inferences/inference.py [hunk 5]
```diff
@@ -258,3 +284,48 @@ def finalize(self):
...
+
+     Returns
+     -------
+     None
+
+    """
+    if logvars is None:
+      logvars = []
+      for k in self.latent_vars:
+        logvars += get_variables(self.latent_vars[k])
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: variables can also exist as model parameters. I recommend an additional loop applying this to keys of `self.data` and the priors (which are keys of `self.latent_vars`).

File: edward/version.py [hunk 6]
```diff
@@ -1 +1 @@
-__version__ = '1.3.1'
+__version__ = '1.3.1-closedloop'
```
Comment (dustinvtran) [inline edward/version.py]: Can remove this :)

File: examples/normal_sghmc.py [hunk 7]
```diff
@@ -50,6 +50,7 @@ def mvn_plot_contours(z, label=False, ax=None):
 qz = Empirical(params=tf.Variable(tf.random_normal([5000, 2])))
 
 inference = ed.SGHMC({z: qz})
+inference.initialize(logdir='C:\\Users\skruz\\temp\\tensorflow\\')
```
Comment (dustinvtran) [inline examples/normal_sghmc.py]: You can remove this argument to be more generic, say, `logdir='log'`.

**Comment by dustinvtran** [comment]
Wow, this is an amazing PR @closedLoop. It's a very valuable contribution in making Edward fits easier to explore and also to diagnose model/learning problems. And thanks for the demo!

I added comments above. Two others

+ Can you also remove changes from the various notebooks unrelated to this PR?
+ I recommend displaying the TensorBoard pics in the tutorial. You can do that by using
```python
![GAN Fig 0](https://raw.githubusercontent.com/blei-lab/edward/master/docs/images/gan-fig0.png)
```
Then you can explain them as you do well in the video demo.

File: edward/inferences/inference.py (hunk 1 above)
Comment (dustinvtran) [inline edward/inferences/inference.py]: Also: if logrun can easily be specified as part of logdir that would also enable the timestampping for those who prefer to control inference using `initialize()` then `updates()`. (that is, they don't call `run()`)

File: edward/inferences/inference.py [hunk 8]
```diff
@@ -258,3 +284,48 @@ def finalize(self):
...
+      # If variable is a one dimensional tensor, log each element in the tensor
+      # individually. Only log the first log_max_scalers_per_var variables
+      if len(var.shape) == 1:
+        for i in range(var.shape[0]):
+          if log_max_scalers_per_var is None or i < log_max_scalers_per_var:
+            tf.summary.scalar('{}/{}'.format(var_name, i), var[i])
+
+      # If var is multi-dimensional, log the distribution
+      if len(var.shape) > 0 and np.max(var.shape) > 1:
+        tf.summary.histogram(var_name, var)
```
Comment (dustinvtran) [inline edward/inferences/inference.py]: Some of this conflicts with the current tensorboard tracking in `variational_inference.py` (see #654). My guess is you can remove the parameter histograms in `variational_inference.py`. 

I also recommend checking to see how the naming systems match.

File: edward/inferences/inference.py (hunk 1 above)
Comment (closedLoop) [inline edward/inferences/inference.py]: They way I thought about it is that `logrun` will tend to be pretty static for a given project and you want a way to cleanly separate info each time you refit your models.  Therefor you can leave it blank as you arbitrarily refine the model and use the timestamps to track your progress.  Else you could set `logrun` to something like `using_adam` vs `using_adagrad` to highlight in TensorBoard the convergence results of each.

I wanted to keep `logdir` a string since it is a string in TensorBoard, and given it has the same name I wanted to make the usage similar to avoid confusion.

File: edward/inferences/inference.py (hunk 2 above)
Comment (closedLoop) [inline edward/inferences/inference.py]: The T was a means of just separating the date and the time segments like most of the conventions here [https://www.loggly.com/docs/automated-parsing/#system](https://www.loggly.com/docs/automated-parsing/#system) and also removing all of the other non-numeric characters from the conventions.  However after using this in tensorboard a bit.  I'm going to replace the 'T' with a '_' just to keep it easier to read on the eye.

File: edward/inferences/inference.py (hunk 1 above)
Comment (dustinvtran) [inline edward/inferences/inference.py]: Users can already separate info each time they refit a model by tweaking the `logdir` argument.  A typical use case would be
```python
inference.initialize(logdir='log/lr_{}_nsamples_{}'.format(learning_rate_stepsize, n_samples))
```
With `logdir`, users have full flexibility in setting however many subdirectories and filenames they want. `logdir + logrun` sort of imposes structure, where a user can specify `logrun` already inside `logdir`.

That said, I love the idea of having a default subdirectory inside `logdir` which timestamps the run. It seems impossible to turn it on/off without having an additional argument. So then maybe `logrun` should just be a boolean, and its name+default be changed to something like `log_timestamp=True`?

**Comment by closedLoop** [comment]
@dustinvtran this has been cleaned up integrated with the gradient and loss tracking from the merged pull request #598 .  I think it's looking really good and I've tested it on some larger models and it works well there too.

Main changes:
 * parameter naming is consistent with logging in ``variational_inference.py``
 * ``logrun`` has been changed to ``log_timestamp`` boolean based on your suggestion.
 * I've removed ``log_max_scalars_per_var`` from my previous addition.  It was unnecessarily complicated and was a solution in search of a problem.  
 * Single valued ``Variables`` are graphed in ``Scalars`` else they are logged to ``Histogram`` and ``Distributions`` in TensorBoard. 
 * The ``tensorboard.ipynb`` has been updated and includes references to the gradient tracking and loss logged that was added in #598 

Thank you for all of your suggestions.  I believe this is much cleaner now.  I'm happy to modify anything else you'd like

Here's the new screenshots!
tensorboard_scalars

tensorboard_histograms

**Comment by dustinvtran** [comment]
Great work. Merging now. I'll add a separate PR that adds the equivalent notebook as a tutorial on the website.
